akka-actor/src/main/scala/akka/pattern/Patterns.scala

```diff
@@ -485,6 +485,56 @@ object Patterns {
       scheduler: Scheduler,
       ec: ExecutionContext): CompletionStage[T] =
     scalaRetry(() => attempt.call().toScala, attempts, delay.asScala)(ec, scheduler).toJava
+
+  /**
+   * Returns an internally retrying [[scala.concurrent.Future]].
+   * The first attempt will be made immediately, and each subsequent attempt will be made after 'delay'.
+   * A scheduler (eg context.system.scheduler) must be provided to delay each retry.
+   * You could provide a function to generate the next delay duration after first attempt,
+   * this function should never return `null`, otherwise an [[IllegalArgumentException]] will be through.
+   * If attempts are exhausted the returned future is simply the result of invoking attempt.
+   * Note that the attempt function will be invoked on the given execution context for subsequent tries and
+   * therefore must be thread safe (not touch unsafe mutable state).
+   */
+  def retry[T](
+      attempt: Callable[Future[T]],
+      attempts: Int,
+      initialDelay: FiniteDuration,
+      nextDelayFunction: akka.japi.function.Function3[Int, FiniteDuration, FiniteDuration, FiniteDuration],
```

Comment:
Do we need 2 `FiniteDuration` parameters here? I'd say 1 would be sufficient, or perhaps it would be enough to just take a `Int => FiniteDuration` that returns the delay for the `n`th attempt.

Then we can also drop the `initialDelay` since that is just the result of this function when called with `0`.

Also, we could make it `Int => Option[FiniteDuration]`, so `None` would signal that there are no retries left. That would make it possible to have a dynamic number of retries, and remove the need for the `attempts` parameter.

Reply:
I will take a look at this tonight

Reply:
I have updated it again.